### DISCUSSION TITLE
plugin backends

### DISCUSSION THREAD

**Discussion by jdx** [jdx/mise#3709] [Ideas]
we should have the ability for plugins to define backends. Right now they have to be specified in the core rust code.

This will likely involve a challenging refactor as plugins are currently directly tied to the asdf backend for obvious reasons. A major part of why inside the codebase the term "forge" was adopted was to make this transition easier since "plugin" no longer makes sense when plugins can define their own "forges".

**Comment by iilyak** [comment]
If plugins mechanism is not yet present the following looks like an interesting approach https://extism.org/.

There is also implementation of plugins in typst https://github.com/typst/typst/blob/56ecd6c806ab32a7cca5289500504e398198d283/crates/typst/src/foundations/plugin.rs#L120

**Comment by jdx** [comment]
it's worth considering. Extism is a great framework. That said, mise already has a very mature plugin model that could easily be extended. One benefit of extism would be windows support though.

**Comment by jdx** [comment]
I'm not so sure how useful this would be. I think the reality is I have no need for it since I have the ability to make backends and I don't know how popular plugins will be in the post-asdf world of mise. I think people use mise for the default tools mostly.

I may be wrong and would be happy to reopen if someone has a use-case here. Right now though this hasn't been something a single person has asked for (at least that I can recall) so I'm going to close but will consider reopening if I hear a use-case—at least a use-case I like

A few people thumbed this one though, so I'll tag them

**Comment by jdx** [comment]
@nickchomey @shotanue @bryan-hoang @bhanutejags @iilyak

**Comment by jdx** [comment]
there was a use-case mentioned in discord so I'll reopen, the proposed idea was a potential mise-gcloud plugin could be used to provide phony components:

```toml
[tools]
"gcloud" = "latest"
"gcloud:foo" = "latest"
```

while this is possible today with postinstall hooks, that has issues like how the tool options are not run if the tool is already installed without them—possibly in a different config.

That feature on it's own is probably not worth the huge lift this functionality would require to support, but it does make me think this ticket at least should stay open. I suspect there are other use-cases waiting to be uncovered.

**Comment by andrewcrook** [comment]
> it's worth considering. Extism is a great framework. That said, mise already has a very mature plugin model that could easily be extended. One benefit of extism would be windows support though.

Also I would love to develop some ideas for Mise, however, Rust isn't my daily driver so language agnostic would be great.
I have experimented using different languages with ASDF plugins. 

Anyway,  I think distributing compiled executables (and/or  WASM) is a good idea you wouldn’t want to increase the number dependencies just to use a plugin. So self contained statically compiled and/or WASM makes sense for an offical repository. I believe you already do this with Mise native plugins which are written in Rust.

Most scripting languages have options to make self contained executables or WASM, JVM based like Clojure, Kotlin and Java have GraalVM native images, GraalVM also has support for other Languages, LLVM has some interesting work and of course Rust, Zig, C/C++, Go, Swift, Mono C# etc are all compiled by default. 

@jdx Extism sounds interesting what kind of IPC does it use locally?

**Comment by jdx** [comment]
what is it you're looking to do? part of the reason I haven't started on this is I haven't found many use-cases that would need it. gcloud is one, but it's pretty minor.

**Comment by ejrichards** [comment]
Just to follow up on https://github.com/jdx/mise/pull/4704 and https://github.com/jdx/mise/discussions/3538#discussioncomment-12363091

I wouldn't say it is the most crucial feature, but it would allow mise to integrate with the Nix / NixOS ecosystem.  In my mind it would fill a small gap where you don't want to write or maintain a whole flake to define the dev environment or you are on a team that only has some people using Nix / NixOS, so you can just maintain mise as the "source of truth" for the dev environment.

**Comment by jdx** [comment]
https://mise.jdx.dev/backend-plugin-development.html
